Sources/SwiftProtobuf/JSONScanner.swift

```diff
@@ -79,41 +79,143 @@ private func fromHexDigit(_ c: UnicodeScalar) -> UInt32? {
+/* 0x10 */ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+/* 0x20 */ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, -1, -1, 63,
+/* 0x30 */ 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, -1, -1, -1, -1, -1, -1,
+/* 0x40 */ -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
+/* 0x50 */ 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1,
+/* 0x60 */ -1, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
+/* 0x70 */ 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, -1, -1, -1, -1, -1,
+/* 0x80 */ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+/* 0x90 */ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+/* 0xa0 */ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+/* 0xb0 */ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+/* 0xc0 */ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+/* 0xd0 */ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+/* 0xe0 */ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+/* 0xf0 */ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+]
+
 /// Returns a `Data` value containing bytes equivalent to the given
 /// Base64-encoded string, or nil if the conversion fails.
-private func decodeBytes(base64String s: String) -> Data? {
-  var out = [UInt8]()
-  let digits = s.utf8
-  var n = 0
-  var bits = 0
-  for (i, digit) in digits.enumerated() {
-    n <<= 6
-    switch digit {
-    case asciiUpperA...asciiUpperZ: n |= Int(digit - asciiUpperA); bits += 6
-    case asciiLowerA...asciiLowerZ:
-        n |= Int(digit - asciiLowerA + UInt8(26)); bits += 6
-    case asciiZero...asciiNine: n |= Int(digit - asciiZero + UInt8(52)); bits += 6
-    case 43: n |= 62; bits += 6
-    case 47: n |= 63; bits += 6
-    case 61: n |= 0
-    default:
-      return nil
+///
+/// Notes on Google's implementation:
+///  * Google's C++ implementation accepts arbitrary whitespace
+///    mixed in with the base-64 characters
+///  * Google's C++ implementation ignores missing '=' characters
+///    but if present, there must be the exact correct number of them.
+///  * Google's C++ implementation accepts both "regular" and
+///    "web-safe" base-64 variants (it seems to prefer the
+///    web-safe version as defined in RFC 4648
+private func parseBytes(
+  source: UnsafeBufferPointer<UInt8>,
+  index: inout UnsafeBufferPointer<UInt8>.Index,
+  end: UnsafeBufferPointer<UInt8>.Index
+) throws -> Data {
+    let c = source[index]
+    if c != asciiDoubleQuote {
+        throw JSONDecodingError.malformedString
     }
-    if i % 4 == 3 {
-      out.append(UInt8(truncatingBitPattern: n >> 16))
-      if bits >= 16 {
-        out.append(UInt8(truncatingBitPattern: n >> 8))
-        if bits >= 24 {
-          out.append(UInt8(truncatingBitPattern: n))
+    source.formIndex(after: &index)
+
+    // Count the base-64 digits
```

Comment:
Good point.  I should experiment.  Because you would have to find the close quote first, your idea would boil down to:
* Scan for the close quote (faster than my current first pass)
* Size the output to 3/4 of that
* Parse
* Truncate if necessary (probably not, assuming whitespace is uncommon)
This might also be marginally simpler than my current code, which would be another plus.